cache/redis_client.go

```diff
@@ -0,0 +1,347 @@
+	opts := &redis.UniversalOptions{
+		Addrs:        strings.Split(config.Endpoint.String(), ","),
+		Username:     config.Username,
+		Password:     config.Password.String(),
+		DB:           config.DB,
+		MasterName:   config.MasterName,
+		DialTimeout:  config.DialTimeout,
+		ReadTimeout:  config.ReadTimeout,
+		WriteTimeout: config.WriteTimeout,
+		PoolSize:     config.ConnectionPoolSize,
+		MinIdleConns: config.MinIdleConnections,
+		MaxConnAge:   config.MaxConnectionAge,
+		IdleTimeout:  config.IdleTimeout,
+	}
+
+	if config.TLSEnabled {
+		tlsConfig := config.TLS
+
+		tlsClientConfig, err := newClientConfig(logger, tlsConfig.CertPath, tlsConfig.KeyPath,
+			tlsConfig.CAPath, tlsConfig.ServerName, tlsConfig.InsecureSkipVerify)
+
+		if err != nil {
+			return nil, err
+		}
+
+		opts.TLSConfig = tlsClientConfig
+	}
+
+	cli := redis.NewUniversalClient(opts)
+	if reg != nil {
+		reg = prometheus.WrapRegistererWith(prometheus.Labels{"name": name}, reg)
+	}
+
+	c := &redisClient{
+		UniversalClient: cli,
+		config:          config,
+		logger:          logger,
+	}
+	if config.MaxGetMultiConcurrency > 0 {
+		c.getMultiGate = gate.New(
+			prometheus.WrapRegistererWithPrefix("redis_getmulti_", reg),
+			config.MaxGetMultiConcurrency,
+		)
+	}
+
+	c.metrics = newClientMetrics(
+		prometheus.WrapRegistererWithPrefix("redis_", reg),
+	)
+
+	c.asyncQueue = newAsyncQueue(config.MaxAsyncBufferSize, config.MaxAsyncConcurrency)
+
+	return c, nil
+}
+
+// SetAsync implement RemoteCacheClient.
+func (c *redisClient) SetAsync(ctx context.Context, key string, value []byte, ttl time.Duration) error {
+	// Skip hitting redis at all if the item is bigger than the max allowed size.
+	if c.config.MaxItemSize > 0 && uint64(len(value)) > uint64(c.config.MaxItemSize) {
+		c.metrics.skipped.WithLabelValues(opSet, reasonMaxItemSize).Inc()
+		return nil
+	}
+
+	err := c.asyncQueue.submit(func() {
+		start := time.Now()
+
+		_, err := c.Set(ctx, key, value, ttl).Result()
+		if err != nil {
+			level.Debug(c.logger).Log(
+				"msg", "failed to store item to redis",
+				"key", key,
+				"sizeBytes", len(value),
+				"err", err,
+			)
+			c.trackError(opSet, err)
+			return
+		}
+
+		c.metrics.dataSize.WithLabelValues(opSet).Observe(float64(len(value)))
+		c.metrics.duration.WithLabelValues(opSet).Observe(time.Since(start).Seconds())
+	})
+
+	if errors.Is(err, errAsyncQueueFull) {
+		c.metrics.skipped.WithLabelValues(opSet, reasonAsyncBufferFull).Inc()
+		level.Debug(c.logger).Log("msg", "failed to store item to redis because the async buffer is full", "err", err, "size", c.config.MaxAsyncBufferSize)
+		return nil
+	}
+	return err
+}
+
+// GetMulti implement RemoteCacheClient.
+func (c *redisClient) GetMulti(ctx context.Context, keys []string) map[string][]byte {
+	if len(keys) == 0 {
+		return nil
+	}
+	start := time.Now()
+	results := make(map[string][]byte, len(keys))
+	var mu sync.Mutex
+	var total int
```

Comment:
[nit] Just for clarity.

```suggestion
	var cacheHitBytes int
```